third_party/move/move-vm/types/src/values/values_impl.rs

```diff
@@ -3710,19 +3744,34 @@ impl<'c, 'l, 'v> serde::Serialize
 
             // Signer.
             (L::Signer, ValueImpl::Container(Container::Struct(r))) => {
-                let v = r.borrow();
-                if v.len() != 1 {
-                    return Err(invariant_violation::<S>(format!(
-                        "cannot serialize container as a signer -- expected 1 field got {}",
-                        v.len()
-                    )));
+                if self.ctx.legacy_signer {
+                    // Only allow serialization of master signer.
+                    if *r.borrow()[0].as_value_ref::<u16>().map_err(|_| {
+                        invariant_violation::<S>(format!(
+                            "First field of a signer needs to be an enum descriminator, got {:?}",
+                            self.value
+                        ))
+                    })? != 0
+                    {
+                        return Err(S::Error::custom(PartialVMError::new(StatusCode::ABORTED)));
```

Comment:
Adde  error message?

Reply:
The error msg wouldn't help here that much, as the on chain serializer will convert this error into an error code, and we don't have a good way of passing that around.